src/lib/gossip_net/gossip_net.ml

```diff
@@ -218,7 +222,10 @@ module Make (Message : Message_intf) :
                      !"Peer %{sexp: Unix.Inet_addr.t} banned, disconnecting."
                      addr ;
                    Deferred.List.iter (Hashtbl.to_alist conn_list)
-                     ~f:(fun (_, conn) -> Rpc.Connection.close conn ) )) ;
+                     ~f:(fun (_, conn_opt) ->
+                       match conn_opt with
+                       | None -> return ()
```

Comment:
There's another race here, if we ban a peer and there is a connection from that peer that hasn't called `connection_state` yet. This code should wait until there is a connection object to close. Best way to do that is probably to change the `Rpc.Connection.t option` to `Rpc.Connection.t Ivar.t`, put an empty Ivar in where you had `None` before, fill it where you set the option to `Some` and use `Ivar.read` here.